lib/google/gax/credentials.rb

```diff
@@ -0,0 +1,149 @@
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+require 'forwardable'
+require 'googleauth'
+require 'json'
+require 'signet/oauth_2/client'
+
+module Google
+  module Gax
+    # @private
+    # Represents the OAuth 2.0 signing logic.
+    # This class is intended to be inherited by API-specific classes
+    # which overrides the SCOPE constant.
+    class Credentials
+      TOKEN_CREDENTIAL_URI = 'https://accounts.google.com/o/oauth2/token'.freeze
+      AUDIENCE = 'https://accounts.google.com/o/oauth2/token'.freeze
+      SCOPE = [].freeze
+      PATH_ENV_VARS = [].freeze
+      JSON_ENV_VARS = [].freeze
+      DEFAULT_PATHS = [].freeze
+
+      attr_accessor :client
+
+      ##
+      # Delegate client methods to the client object.
+      extend Forwardable
+      def_delegators :@client,
+                     :token_credential_uri, :audience,
+                     :scope, :issuer, :signing_key, :updater_proc
+
+      def initialize(keyfile, scope: nil)
+        verify_keyfile_provided! keyfile
+        if keyfile.is_a? Signet::OAuth2::Client
+          @client = keyfile
+        elsif keyfile.is_a? Hash
+          hash = stringify_hash_keys keyfile
+          hash['scope'] ||= scope
+          @client = init_client hash
+        else
+          verify_keyfile_exists! keyfile
+          json = JSON.parse ::File.read(keyfile)
+          json['scope'] ||= scope
+          @client = init_client json
+        end
+        @client.fetch_access_token!
+      end
+
+      ##
+      # Returns the default credentials.
+      #
+      def self.default(scope: nil)
+        env = ->(v) { ENV[v] }
+        json = lambda do |v|
+          unless ENV[v].nil?
+            begin
+              JSON.parse ENV[v]
+            rescue
+              nil
+            end
+          end
+        end
+        path = ->(p) { ::File.file? p }
+
+        # First try to find keyfile file from environment variables.
+        self::PATH_ENV_VARS.map(&env).compact.select(&path)
+                           .each do |file|
+          return new file, scope: scope
+        end
+        # Second try to find keyfile json from environment variables.
+        self::JSON_ENV_VARS.map(&json).compact.each do |hash|
+          return new hash, scope: scope
+        end
+        # Third try to find keyfile file from known file paths.
+        self::DEFAULT_PATHS.select(&path).each do |file|
+          return new file, scope: scope
+        end
+        # Finally get instantiated client from Google::Auth.
+        scope ||= self::SCOPE
+        client = Google::Auth.get_application_default scope
+        new client
+      end
+
+      protected
+
+      # Verify that the keyfile argument is provided.
+      def verify_keyfile_provided!(keyfile)
+        raise 'You must provide a keyfile to connect with.' if keyfile.nil?
```

Comment:
When could this get raised? It seems wrong to tell users they "must" provide a keyfile given that for GAPIC, we expect most people to use ADC, and the keyfile to be a more advanced use case.